File: rubygems/pray-cli/lib/pray/resolve_tarball.rb
# frozen_string_literal: true

require "fileutils"
require "pathname"
require_relative "archive_unpack"
require_relative "hashing"
require_relative "resource_limits"

module Pray
  module ResolveTarball
    module_function

    def package_root(project_root, tarball, offline:)
      artifact_bytes = read_tarball_bytes(project_root, tarball, offline: offline)
      if artifact_bytes.bytesize > ResourceLimits::MAX_ARCHIVE_TOTAL_BYTES
        raise Error.integrity(
          "package archive exceeds #{ResourceLimits::MAX_ARCHIVE_TOTAL_BYTES} bytes"
        )
      end

      cache_key = Hashing.sha256_prefixed(artifact_bytes).delete_prefix("sha256:")[0, 16]
      cache_directory = File.join(project_root, ".pray", "cache", "tarball", cache_key)
      if File.directory?(cache_directory)
        begin
          Resolve.find_prayspec_file(cache_directory)
          return cache_directory
        rescue Error
          nil
        end
      end

      staging_directory = "#{cache_directory}.staging"
      FileUtils.rm_rf(staging_directory)
      FileUtils.mkdir_p(staging_directory)
      ArchiveUnpack.unpack_praypkg(artifact_bytes, staging_directory)
      Resolve.find_prayspec_file(staging_directory)
      FileUtils.mkdir_p(File.dirname(cache_directory))
      FileUtils.rm_rf(cache_directory)
      FileUtils.mv(staging_directory, cache_directory)
      cache_directory
    end

    def read_tarball_bytes(project_root, tarball, offline:)
      if tarball.start_with?("http://", "https://")
        if offline
          raise Error.resolution(
            "tarball #{tarball} is not cached locally and offline mode is enabled"
          )
        end
        return Registry.http_get(tarball)
      end

      path = local_tarball_path(project_root, tarball)
      unless File.file?(path)
        raise Error.resolution("tarball missing at #{path}")
      end

      File.binread(path)
    end

    def local_tarball_path(project_root, tarball)
      path = tarball.delete_prefix("file://")
      Pathname.new(path).absolute? ? path : File.expand_path(path, project_root)
    end
  end
end

File: crates/pray-core/src/resolve_tarball.rs
use crate::hashing::sha256_prefixed;
use crate::package_archive::unpack_praypkg;
use crate::paths::{find_prayspec_file, remove_path_if_exists};
use crate::resolve_context::ResolveOptions;
use crate::resource_limits::MAX_ARCHIVE_TOTAL_BYTES;
use crate::{PrayError, PrayResult};
use std::fs;
use std::path::{Path, PathBuf};

pub(crate) fn resolve_tarball_package_root(
    project_root: &Path,
    tarball: &str,
    options: &ResolveOptions,
) -> PrayResult<PathBuf> {
    let artifact_bytes = read_tarball_bytes(project_root, tarball, options)?;
    if artifact_bytes.len() as u64 > MAX_ARCHIVE_TOTAL_BYTES {
        return Err(PrayError::Integrity(format!(
            "package archive exceeds {MAX_ARCHIVE_TOTAL_BYTES} bytes"
        )));
    }
    let cache_key = sha256_prefixed(&artifact_bytes)
        .trim_start_matches("sha256:")
        .chars()
        .take(16)
        .collect::<String>();
    let cache_directory = project_root.join(".pray/cache/tarball").join(cache_key);
    if find_prayspec_file(&cache_directory).is_ok() {
        return Ok(cache_directory);
    }
    let staging_directory = cache_directory.with_extension("staging");
    remove_path_if_exists(&staging_directory)?;
    fs::create_dir_all(&staging_directory)?;
    if let Err(error) = unpack_praypkg(&artifact_bytes, &staging_directory) {
        let _ = remove_path_if_exists(&staging_directory);
        return Err(error);
    }
    if let Err(error) = find_prayspec_file(&staging_directory) {
        let _ = remove_path_if_exists(&staging_directory);
        return Err(error);
    }
    if let Some(parent) = cache_directory.parent() {
        fs::create_dir_all(parent)?;
    }
    remove_path_if_exists(&cache_directory)?;
    fs::rename(&staging_directory, &cache_directory).map_err(|error| {
        let _ = remove_path_if_exists(&staging_directory);
        PrayError::from(error)
    })?;
    Ok(cache_directory)
}

fn read_tarball_bytes(
    project_root: &Path,
    tarball: &str,
    options: &ResolveOptions,
) -> PrayResult<Vec<u8>> {
    if tarball.starts_with("http://") || tarball.starts_with("https://") {
        if options.offline {
            return Err(PrayError::Resolution(format!(
                "tarball {tarball} is not cached locally and offline mode is enabled"
            )));
        }
        return crate::registry_http::http_get(tarball);
    }
    let path = local_tarball_path(project_root, tarball);
    fs::read(&path).map_err(|error| {
        PrayError::Resolution(format!("tarball missing at {}: {error}", path.display()))
    })
}

fn local_tarball_path(project_root: &Path, tarball: &str) -> PathBuf {
    let path = tarball.strip_prefix("file://").unwrap_or(tarball);
    let path = Path::new(path);
    if path.is_absolute() {
        path.to_path_buf()
    } else {
        project_root.join(path)
    }
}
